[WIP] Minimax search function

a1shadows opened the pull request:
Issue #9

a1shadows committed:
Game trait created.

a1shadows committed:
parameter error resolved

a1shadows committed:
removed implicit parameter

a1shadows committed:
added missing isTerminalState function

a1shadows committed:
minimaxsearch added

a1shadows commented:
@BusyByte Hey, could you please resolve the conflicts and review the pull request. Thanks.

BusyByte committed:
Merge branch 'master' into master

BusyByte commented on core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala:
```diff
@@ -0,0 +1,44 @@
+  */
+
+object MinimaxDecision {
+    def minMaxDecision[S, A, P](g: Game[S][A][P]): S, P => A = {
```
I'm not a big fan of single letter parametric names when they actually mean things.  I'd use upper case and use the full name like `STATE`, `ACTION`, and `PERCEPT` rather than `S`, `A`, and `P` respectively.

BusyByte commented on core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala:
```diff
@@ -0,0 +1,44 @@
+
+        def minMaxValue(actions: List[Action], p: P): Action = actions match {
+            case Nil => NoAction
+            case List(a: Action) => a
```
see above

BusyByte commented on core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala:
```diff
@@ -0,0 +1,44 @@
+            case a1 :: a2 :: rest => maxMinValue( (if (g.minValue(a1, p) > g.minValue(a2, p)) a1 else a2) :: rest )
+        }
+
+        def minMaxValue(actions: List[Action], p: P): Action = actions match {
```
It appears that this may be tail recursive.  please add a `@tailrec` annotation to the method to checkit is.

BusyByte commented on core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala:
```diff
@@ -0,0 +1,44 @@
+    def minMaxDecision[S, A, P](g: Game[S][A][P]): S, P => A = {
+        (s: S, p: P) => maxMinValue(g.getActions(s), p)
+            
+        def maxMinValue(actions: List[Action], p: P): Action = actions match {
```
Rather than using `Action` should be using the parametric type `ACTION` (you currently have as `A`)

BusyByte commented on core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala:
```diff
@@ -0,0 +1,44 @@
+        (s: S, p: P) => maxMinValue(g.getActions(s), p)
+            
+        def maxMinValue(actions: List[Action], p: P): Action = actions match {
+            case Nil => NoAction
```
You'll not want to use `NoAction` but instead pass into minMaxDecision a parameter of type `ACTION` (you currently have as A) 

eg
`def minMaxDecision[S, A, P](g: Game[S][A][P], noAction: A)`

BusyByte reviewed (changes requested):
Some changes needed but looks good overall.
Can you add a test specification (specs2) and/or a scalacheck (property based check) to prove it works?
If there are examples from the book sometimes I'll use those to prove it works.

BusyByte commented:
@a1shadows sorry something wasn't quite right with my resolution of the merge conflicts, hopefully you can fix it up

BusyByte commented:
@a1shadows let me know if you have any questions on the above or need help

a1shadows committed:
1st requested change

case a :: Nil used instead to improve consistency with other cases.

a1shadows committed:
2nd, 3rd, 4th, 5th, 6th requested changes made, errors resolved

Integrated all requested changes and remedied handful of syntactical and logical errors.

a1shadows commented:
@BusyByte Have tried addressing all the requested changes. Am working on testing the code. Could you please review the changes?

BusyByte reviewed (approved):
@a1shadows this looks good but still missing a test specification/property based test.

--edit never mind above, I just saw you were working on the testing code still

a1shadows commented:
> @a1shadows this looks good but still missing a test specification/property based test

I'll do the test code. I have my mid-terms till the 28th. Just need a little bit of time. Thanks for your constant guidance.

a1shadows committed:
added test problem for minimax

a1shadows committed:
test added

test spec for minimax

a1shadows commented:
@BusyByte I've written a test for the function, but I'm having some difficulty coming up with more tests. Could you please review the new code and suggest how I should proceed with the tests?

BusyByte commented:
@a1shadows fyi travis has broken, I've merge a fix to master for it now:
https://github.com/aimacode/aima-scala/pull/72

you'll need to merge master in order to get travis to run your tests

BusyByte commented on core/src/main/scala/aima/core/search/problems/TwoPlyGame.scala:
```diff
@@ -0,0 +1,60 @@
+/**
+  * @author Aditya Lahiri
+  */
+case class Player(name: String)     extends AnyVal
```
if there can only be 2 players why not code it that way;
```scala
sealed abstract class Player {
def name: String
}

final case class Player1(name: String) extends Player
final case class Player2(name: String) extends Player
```

BusyByte commented on core/src/main/scala/aima/core/search/problems/TwoPlyGame.scala:
```diff
@@ -0,0 +1,60 @@
+                         1 -> List(Action(4), Action(5), Action(6)),
+                         2 -> List(Action(7), Action(8), Action(9)),
+                         3 -> List(Action(10), Action(11), Action(12)))
+  val States = Map(0 -> State(0),
```
Seems like this might go away with changes I outlined below to Action.

a1shadows committed:
Merge pull request #1 from aimacode/master

catching up

BusyByte commented:
@a1shadows might be worth going through a full game and verifying each player's actions are what you expect and not just the first one and verify you get NoAction at the end.
I left some suggestions above but this seems pretty good so far.

BusyByte commented:
@a1shadows also since this is `adversarial` maybe verify who should `win` and maybe a comment as to why

a1shadows committed:
Merge branch 'master' into pr/2

a1shadows committed:
Merge pull request #3 from a1shadows/pr/2

Pr/2

a1shadows commented:
I'm getting a ton of errors on the code when I'm running it on the sbt which weren't there earlier. I can't understand them.
They are all of this kind:

aima-scala\core\src\main\scala\aima\core\agent\Actuator.scala:8:16: type parameter Action defined in trait Actuator shadows trait Action defined in package agent. You may want to rename your type parameter, or possibly remove it.

Could you point me to what's going wrong, please?

a1shadows commented:
Also, would it be possible to set up a gitter channel to discuss the repository? For general queries and the like?

BusyByte commented:
@a1shadows I've created a gitter channel, you should get an invitation,  I think the error you are seeing is because of a name collision.  Maybe try upper case `A` in your type parameter or upper case `ACTION`
for example.
`trait Foo[A]` or `trait Bar[ACTION]`

BusyByte commented:
@a1shadows maybe you need to do a sbt clean ` sbt ";clean;test:compile"` after you merge master into your branch.
I don't think there is a `trait Action defined in package agent` in package agent anymore looking at master.

BusyByte commented:
@a1shadows this still says `WIP` in the title.  Is there anything else you'd like to do before merging this?

a1shadows commented:
Yeah. I'd like to work a bit on it.

On Sat, 29 Jun 2019, 5:00 pm Shawn Garner, <notifications@github.com> wrote:

> @a1shadows <https://github.com/a1shadows> this still says WIP in the
> title. Is there anything else you'd like to do before merging this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aimacode/aima-scala/pull/71?email_source=notifications&email_token=AFIGZIQM36IBESICN2TJWGDP45BWHA5CNFSM4HCLQQRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY3XFGA#issuecomment-506950296>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFIGZISJXT33CFT5IDIL6M3P45BWHANCNFSM4HCLQQRA>
> .
>

a1shadows commented:
@BusyByte Hey, I was attempting to rewrite the code to give an optimal state for the game at every level as opposed to my earlier code that only gave the next best move, but I just realized that the pseudocode only provides an algorithm for the next best move with the assumption that the "MAX" player is making the move. I am unsure how to proceed from here on. Should I try to write the code for the complete algorithm that can calculate moves for both "MIN" and "MAX" players or should I restrict myself to what the book has provided?

a1shadows commented on core/src/main/scala/aima/core/search/problems/TwoPlyGame.scala:
```diff
@@ -0,0 +1,60 @@
+case class State(stateNumber: Int)  extends AnyVal
+case class Action(stateNumber: Int) extends AnyVal
+
+object TwoPlyGame extends Game[Player, State, Action] {
```
Could you please give an example of the format in which I'm supposed to give the comment about the function?

BusyByte commented:
@a1shadows Looking at the [test for the Java version](https://github.com/aimacode/aima-java/blob/AIMA4e/test/src/test/java/aima/test/unit/search/adversarial/SearchForAdversarialActionFunctionTest.java#L67) it looks like it should handle both players.

Looking at the [java implementation](https://github.com/aimacode/aima-java/blob/AIMA4e/core/src/main/java/aima/core/search/basic/adversarial/MinimaxDecision.java#L118)  it just determines which player's turn it is based on the state.  So when you call the search function you give it a state so it should just handle it naturally.

I think it selects the max utility of the minimum utility assuming the next player will pick the min.  So it doesn't look like the algorithm needs changed to handle both users.  

I think your spec may need to change to handle both users.
It looks like the java spec is implementing aima 3e Fig 5.2

Sorry for the delay in response.  Work has been pretty hectic lately.

a1shadows committed:
reformatting

a1shadows committed:
shanges

changes

BusyByte commented:
@a1shadows is this good to go?  If you are satisfied I can merge it.

a1shadows commented:
Ahh. Sorry. I've been busy with work. I'll wrap it up within this weekend.

On Thu, 16 Jan 2020, 1:09 am Shawn Garner, <notifications@github.com> wrote:

> @a1shadows <https://github.com/a1shadows> is this good to go? If you are
> satisfied I can merge it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aimacode/aima-scala/pull/71?email_source=notifications&email_token=AFIGZIVT26T3YQMRU7P2VN3Q55RABA5CNFSM4HCLQQRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJBRPOI#issuecomment-574822329>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIGZIVP5DUXYEQNCSZFT53Q55RABANCNFSM4HCLQQRA>
> .
>

a1shadows committed:
review changes

a1shadows commented:
Still ironing a few things out. I'll let you know when I'm comfortable with the changes. Is that okay?

BusyByte commented:
@a1shadows yeah, that's fine but please prioritize this as there are some more PRs coming through and I don't want you to have a bunch of merge conflicts to work through since your PR predates them.

a1shadows committed:
errors fixed, spec added

a1shadows committed:
Merge pull request #4 from aimacode/master

catching up

a1shadows committed:
clean

a1shadows commented:
I'm satisfied with my changes. Can you review and merge?

BusyByte commented on core/src/main/scala/aima/core/search/adversarial/Game.scala:
```diff
@@ -1,14 +1,27 @@
 
-final case class UtilityValue(value: Double) extends AnyVal
+final case class UtilityValue(value: Double) extends AnyVal {
+  def <(that: UtilityValue)  = if (this.value < that.value) true else false
```
Really this should be `typeclass`

BusyByte commented on core/src/main/scala/aima/core/search/adversarial/Game.scala:
```diff
@@ -1,14 +1,27 @@
+  def getUtility(state: S): UtilityValue
 }
+
+trait Player {}
```
I don't really like having traits with no defined behavior.
I'd just as soon keep these in the definition of the given game so they can be sealed and compile time checked to be total matched.

BusyByte reviewed (approved):
Some things I'd change but this looks good.

BusyByte commented on core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala:
```diff
@@ -0,0 +1,50 @@
+        maxValue(g.result(state, minMaxValue(g.getActions(state))))
+    }
+
+    maxMinValue(g.getActions(g.initialState))
```
@a1shadows I was looking at this when trying to implement alpha beta search and have a question.  Why are you passing in the `g.initialState` rather than `state` which is passed into the function?  this does not seem to match the algorithm.

a1shadows commented:
I decided to not leave that decision up to the caller because the correct
behavior will change according to the state for which the decision will be
made. For some of the nodes, minMaxValue has to be called for the correct
decision and not maxMindecision. Would me like me to change the logic?
Sorry for the late reply.

On Sun, Mar 1, 2020 at 10:50 PM Shawn Garner <notifications@github.com>
wrote:

> *@BusyByte* commented on this pull request.
> ------------------------------
>
> In core/src/main/scala/aima/core/search/adversarial/MinimaxSearch.scala
> <https://github.com/aimacode/aima-scala/pull/71#discussion_r386124778>:
>
> > +
> +    def maxValue(state: STATE): UtilityValue = {
> +      if (g.isTerminalState(state))
> +        g.getUtility(state)
> +      else
> +        minValue(g.result(state, maxMinValue(g.getActions(state))))
> +    }
> +
> +    def minValue(state: STATE): UtilityValue = {
> +      if (g.isTerminalState(state))
> +        g.getUtility(state)
> +      else
> +        maxValue(g.result(state, minMaxValue(g.getActions(state))))
> +    }
> +
> +    maxMinValue(g.getActions(g.initialState))
>
> @a1shadows <https://github.com/a1shadows> I was looking at this when
> trying to implement alpha beta search and have a question. Why are you
> passing in the g.initialState rather than state which is passed into the
> function? this does not seem to match the algorithm.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aimacode/aima-scala/pull/71?email_source=notifications&email_token=AFIGZIXSKVMEYC4KUHB6NODRFKKMLA5CNFSM4HCLQQRKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCXPAWJQ#pullrequestreview-366873382>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIGZIUA4J4SVH7D6FNHY2DRFKKMLANCNFSM4HCLQQRA>
> .
>
